File: scripts/clara.lua
-- scripts/clara.lua
-- Helper module for Lua scripts executing under Clara supervision.

local clara = {}

-- Simple JSON decoder/encoder fallback if external cjson is not available
local function parse_json(str)
    -- Try cjson first if installed
    local ok, cjson = pcall(require, "cjson")
    if ok then
        return cjson.decode(str)
    end

    -- Try dkjson if installed
    local ok2, dkjson = pcall(require, "dkjson")
    if ok2 then
        return dkjson.decode(str)
    end

    -- Very basic regex-based parser or fallback: return raw string wrapper
    -- In practice, lua-cjson or python bridge can be used.
    error("JSON library (cjson or dkjson) required to parse JSON in Lua. Install lua-cjson via luarocks or homebrew.")
end

local function encode_json(tbl)
    local ok, cjson = pcall(require, "cjson")
    if ok then
        return cjson.encode(tbl)
    end
    local ok2, dkjson = pcall(require, "dkjson")
    if ok2 then
        return dkjson.encode(tbl)
    end
    error("JSON library (cjson or dkjson) required to encode JSON in Lua.")
end

-- Get incoming event data passed from Clara supervisor
function clara.get_event()
    local env_json = os.getenv("CLARA_EVENT_JSON")
    if env_json and #env_json > 0 then
        return parse_json(env_json)
    end

    -- Read from stdin
    local input = io.read("*a")
    if input and #input > 0 then
        return parse_json(input)
    end

    return {}
end

-- Call a Clara MCP tool via the clara CLI
function clara.call_tool(tool_name, args)
    local args_parts = {}
    for k, v in pairs(args or {}) do
        local val_str = type(v) == "table" and encode_json(v) or tostring(v)
        -- Escape single quotes in value
        val_str = val_str:gsub("'", "'\\''")
        table.insert(args_parts, string.format("%s='%s'", k, val_str))
    end

    local cmd = string.format("clara tool call %s %s", tool_name, table.concat(args_parts, " "))
    local handle = io.popen(cmd)
    if not handle then
        error("failed to execute clara tool command: " .. cmd)
    end

    local output = handle:read("*a")
    local success, exit_type, code = handle:close()

    if not success then
        error(string.format("tool %s failed with code %s: %s", tool_name, tostring(code), output))
    end

    -- Parse tool response JSON if possible
    local ok, result = pcall(parse_json, output)
    if ok then
        return result
    end
    return output
end

function clara.run_id()
    return os.getenv("CLARA_RUN_ID") or "unknown"
end

function clara.trigger_id()
    return os.getenv("CLARA_TRIGGER_ID") or "unknown"
end

return clara

File: scripts/clara.ts
// scripts/clara.ts
// Standard TypeScript helper SDK for scripts executing under Clara supervision.
// Works seamlessly with Bun, Node, or Deno.

import type { ClaraToolName, ClaraTools, CloudEvent } from "./clara.d.ts";

export type { ClaraToolName, ClaraTools, CloudEvent };

/**
 * Read the incoming CloudEvent payload passed by the Clara supervisor via stdin or environment variable.
 */
export async function readEvent<T = any>(): Promise<CloudEvent<T>> {
  const envJson = process.env.CLARA_EVENT_JSON;
  if (envJson && envJson.trim().length > 0) {
    return JSON.parse(envJson) as CloudEvent<T>;
  }

  // Read from standard input
  let input = "";
  if (typeof Bun !== "undefined") {
    // Bun-native fast stream read
    input = await Bun.stdin.text();
  } else {
    // Node.js fallback
    input = await new Promise<string>((resolve, reject) => {
      let data = "";
      process.stdin.setEncoding("utf8");
      process.stdin.on("data", (chunk) => {
        data += chunk;
      });
      process.stdin.on("end", () => resolve(data));
      process.stdin.on("error", (err) => reject(err));
    });
  }

  if (input && input.trim().length > 0) {
    return JSON.parse(input) as CloudEvent<T>;
  }

  return {
    specversion: "1.0",
    id: "unknown",
    source: "stdin",
    type: "unknown",
    data: {} as T,
  };
}

/**
 * Invokes a Clara MCP tool via the Clara CLI with type-safe arguments and result shapes.
 */
export async function callTool<K extends ClaraToolName>(
  name: K,
  args: ClaraTools[K]["args"]
): Promise<ClaraTools[K]["result"]> {
  const cliArgs: string[] = ["tool", "call", name];

  if (args && typeof args === "object") {
    for (const [k, v] of Object.entries(args)) {
      if (v === undefined) continue;
      const valStr = typeof v === "object" ? JSON.stringify(v) : String(v);
      cliArgs.push(`${k}=${valStr}`);
    }
  }

  let stdout = "";
  let stderr = "";
  let exitCode = 0;

  if (typeof Bun !== "undefined") {
    const proc = Bun.spawn(["clara", ...cliArgs], {
      stdout: "pipe",
      stderr: "pipe",
    });
    stdout = await new Response(proc.stdout).text();
    stderr = await new Response(proc.stderr).text();
    exitCode = await proc.exited;
  } else {
    const { execFile } = await import("node:child_process");
    const { promisify } = await import("node:util");
    const execFileAsync = promisify(execFile);
    try {
      const res = await execFileAsync("clara", cliArgs);
      stdout = res.stdout;
      stderr = res.stderr;
    } catch (err: any) {
      exitCode = err.code ?? 1;
      stdout = err.stdout ?? "";
      stderr = err.stderr ?? err.message;
    }
  }

  if (exitCode !== 0) {
    throw new Error(`Clara tool ${name} failed (exit ${exitCode}): ${stderr || stdout}`);
  }

  try {
    return JSON.parse(stdout);
  } catch {
    return stdout.trim();
  }
}

/**
 * Returns the current Clara execution run ID if running under Clara supervision.
 */
export function runID(): string {
  return process.env.CLARA_RUN_ID || "unknown";
}

/**
 * Returns the current Clara trigger ID if running under Clara supervision.
 */
export function triggerID(): string {
  return process.env.CLARA_TRIGGER_ID || "unknown";
}
